### PR TITLE
Frontend/rest interfaces

### PR DESCRIPTION
**Pull request by gh-phobion**
Implemented BaseRestInterface, AccountRestInterface and SequenceRestInterface

### CODE REVIEW

**Comment by c0derMo** [comment]
Tagging @Kr0nox to review this aswell

File: frontend/restInterfaces/AccountRestInterface.ts
Comment (c0derMo) [inline frontend/restInterfaces/AccountRestInterface.ts]: Small typo, 'successful'

File: frontend/restInterfaces/AccountRestInterface.ts
Comment (c0derMo) [inline frontend/restInterfaces/AccountRestInterface.ts]: Not sure if this works, since we're not actually returning any data, but returning a different status code if login was unsuccessful... I think axios throws an error if a non 200-code is returned, so we could try-catch this and return false in case any error gets thrown, and true if the call passed

File: frontend/restInterfaces/AccountRestInterface.ts
Comment (c0derMo) [inline frontend/restInterfaces/AccountRestInterface.ts]: We dont directly return the number, but rather an object containing the number, so the parameter for the generic should be `<{id: number}>` and add `.id` to the end of the line

File: frontend/restInterfaces/AccountRestInterface.ts
Comment (c0derMo) [inline frontend/restInterfaces/AccountRestInterface.ts]: Likewise here, we need to pass `{ id: accountId }` as a parameter

File: frontend/restInterfaces/AccountRestInterface.ts
Comment (c0derMo) [inline frontend/restInterfaces/AccountRestInterface.ts]: Should use `accountId.join(',')` to convert the number array into the required comma-seperated list

File: frontend/restInterfaces/AccountRestInterface.ts
Comment (c0derMo) [inline frontend/restInterfaces/AccountRestInterface.ts]: Array-Brackets need to be outside of the partial (`Partial<IUser>[]` instead of `Partial<IUser[]>`)

File: frontend/restInterfaces/SequenceRestInterface.ts
Comment (c0derMo) [inline frontend/restInterfaces/SequenceRestInterface.ts]: Same as above, `{ code: string }` instead of `string` inside the generic, and `{ name: title }` as parameter

File: frontend/restInterfaces/SequenceRestInterface.ts
Comment (c0derMo) [inline frontend/restInterfaces/SequenceRestInterface.ts]: `{ code: sequenceCode }` as parameter

File: frontend/restInterfaces/SequenceRestInterface.ts
Comment (c0derMo) [inline frontend/restInterfaces/SequenceRestInterface.ts]: Technically this is an optional extension, so I think this method shouldn't be here for now

File: frontend/restInterfaces/SequenceRestInterface.ts
Comment (c0derMo) [inline frontend/restInterfaces/SequenceRestInterface.ts]: Likewise, this is also an optional extension, so this method should not be here yet

File: frontend/restInterfaces/SequenceRestInterface.ts
Comment (c0derMo) [inline frontend/restInterfaces/SequenceRestInterface.ts]: Last optional extension function that can leave ;)

File: frontend/restInterfaces/AccountRestInterface.ts
Comment (c0derMo) [inline frontend/restInterfaces/AccountRestInterface.ts]: One small addition, this function should get a third parameter called `stayLoggedIn: boolean` which just gets passed on in the request

**Review by c0derMo — CHANGES_REQUESTED** [review]
Good so far, just a couple small changes :)

**Comment by Kr0nox** [comment]
before merging we should diguss how error handling works in the frontend. Do these methods catch the errors or do they get given through to the views?

**Comment by c0derMo** [comment]
> Do these methods catch the errors or do they get given through to the views?

I say let the errors pass on to the views, since they might need to react differently based on different errors - e.g. 404 vs 401, and letting these methods handle errors would make the return types a lot more difficult

**Comment by Kr0nox** [comment]
My suggestion would have been throwing an error with that information. for example that the user did not have the authorization. this would have the advantage that the rest interface is more Abstract and exchangable

**Comment by c0derMo** [comment]
I guess so - should be enough to surround the function calls inside of `BaseRestInterface` with a try-catch and throw new errors (probably `loernwerkErrors`) based on status code?

**Comment by Kr0nox** [comment]
> with a try-catch and throw new errors (probably loernwerkErrors) based on status code?

Will these `loernwerkErrors` have an enum that identifies the error that accoured or will they be written as a string in the message? If we choose the seccond option we would need to define these as public constants to check against somewhere.

File: frontend/restInterfaces/AccountRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/AccountRestInterface.ts]: Parameter name should be plural

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/BaseRestInterface.ts]: Add Error catching

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/BaseRestInterface.ts]: Add Error catching

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/BaseRestInterface.ts]: Add Error catching

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/BaseRestInterface.ts]: Add Error catching

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/BaseRestInterface.ts]: tsdoc for the generic parameter

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/BaseRestInterface.ts]: tsdoc for the generic parameter

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/BaseRestInterface.ts]: tsdoc for the generic parameter

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/BaseRestInterface.ts]: Add Error catching

File: frontend/restInterfaces/SequenceRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/SequenceRestInterface.ts]: I think we should leave this method uninplemented (throw an error that it is not implemented yet), since for it to properly work we would need the factories

File: frontend/restInterfaces/AccountRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/AccountRestInterface.ts]: All mentions of `/account/` should be extracted into a constant

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/BaseRestInterface.ts]: The url needs to be concatinated with the base url (e.g. `http://localhost:5000`). This should be changeble over the vite enviroment.

Comment (Kr0nox) [reply frontend/restInterfaces/BaseRestInterface.ts]: This accounts for all methods in this class

File: frontend/restInterfaces/SequenceRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/SequenceRestInterface.ts]: `/sequence/` should be extracted into a constant in all methods

**Review by Kr0nox — CHANGES_REQUESTED** [review]
Some things I found
I also made comments for the error catching, so we can track whether it is implemented or not

**Comment by c0derMo** [comment]
> Will these `loernwerkErrors` have an enum that identifies the error that accoured or will they be written as a string in the message? If we choose the seccond option we would need to define these as public constants to check against somewhere. 

In the node standards, error codes are usually strings different from the message, we can define them as constants inside `loernwerkUtilities` where the `LoernwerkError` is defined aswell

**Comment by c0derMo** [comment]
Actually, we can kinda do both, if we define an enum like
```
export enum LoernwerkErrorType {
    NOT_FOUND = 'NOT_FOUND'
}
```
inside `loernwerkUtilities` I think

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (c0derMo) [reply frontend/restInterfaces/BaseRestInterface.ts]: We dont even need to make it changable via the vite environment, if we use something like this, which should automatically switch over between local environment and deployed one
`const backendHost = (window.location.hostname.includes('localhost') || window.location.hostname.includes('127.0.0.1')) ? 'http://localhost:5000' : window.location.origin;`

**Comment by Kr0nox** [comment]
Doing both is possible. we would have do decide if we wanna do that

**Comment by c0derMo** [comment]
I prefer enums over "random string constants", so I'd go with that - that way we also comply with the NodeJS standard of string error codes

File: frontend/restInterfaces/AccountRestInterface.ts
Comment (gh-phobion) [reply frontend/restInterfaces/AccountRestInterface.ts]: Do I still need this if I do errorhandling in the BaseRestInterface?

Comment (c0derMo) [reply frontend/restInterfaces/AccountRestInterface.ts]: I'd guess so yeah, since we do want a boolean return value instead of a thrown error - can just catch any error and return false if any error was caught

File: frontend/restInterfaces/SequenceRestInterface.ts
Comment (gh-phobion) [reply frontend/restInterfaces/SequenceRestInterface.ts]: As far as I understood the axios generic methods (e.g. axios.get<T>) should do the job (i.e., converting the plain JSONs to types). Doesn't that work this way? Please refer to: https://upmostly.com/typescript/how-to-use-axios-in-your-typescript-apps for details.

Comment (c0derMo) [reply frontend/restInterfaces/SequenceRestInterface.ts]: Partially yes, but the main issue here is that the things we need to cast are nested within ISequenceWithSlides (`slides.content`), and we need to distinguish which Content gets cast to which ContentType, since we only store `Content`, but need to cast them to `ImageContent`, `TextContent`,... based on their type attributes, which is done by the factories

File: frontend/restInterfaces/AccountRestInterface.ts
Comment (c0derMo) [reply frontend/restInterfaces/AccountRestInterface.ts]: Update: instead of `return await ...` you can just `return true` after the `await ...` call, in case it succeeds, since the endpoint doesn't actually return anything

File: frontend/restInterfaces/SequenceRestInterface.ts
Comment (c0derMo) [inline frontend/restInterfaces/SequenceRestInterface.ts]: Just return `.code` instead of the entire object

File: frontend/restInterfaces/AccountRestInterface.ts
Comment (Kr0nox) [reply frontend/restInterfaces/AccountRestInterface.ts]: > you can just return true after the await ... call

Wouldn't that then throw an error instead of returning `false`?

Comment (c0derMo) [reply frontend/restInterfaces/AccountRestInterface.ts]: If it's
```
try {
  await ...
  return true
} catch {
  return false
}
```
it should return true if the call succeeded, and false otherwise

Comment (Kr0nox) [reply frontend/restInterfaces/AccountRestInterface.ts]: you ment like that. I understood it differently

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (gh-phobion) [reply frontend/restInterfaces/BaseRestInterface.ts]: Lint does not approve if you tag it as @param, so I dont think  it is needed

Comment (Kr0nox) [reply frontend/restInterfaces/BaseRestInterface.ts]: You would use [typeParam](https://typedoc.org/tags/typeParam/) for that

**Comment by Kr0nox** [comment]
@gh-phobion Just noticed your entire files are in the wrong folder. they need to be inside the `src` folder

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (Kr0nox) [inline frontend/restInterfaces/BaseRestInterface.ts]: I think it would be better to transform the statuscodes to more abstarct error mesasges, like NOT_AUTHORIZED

Comment (c0derMo) [reply frontend/restInterfaces/BaseRestInterface.ts]: If we merge main into this branch, we can partially reuse the new LoernwerkErrorCodes enum

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (gh-phobion) [reply frontend/restInterfaces/BaseRestInterface.ts]: lint does not know typeParam

File: frontend/restInterfaces/BaseRestInterface.ts
Comment (Kr0nox) [reply frontend/restInterfaces/BaseRestInterface.ts]: @gh-phobion  Have you checked that your method of assigneg `404` to `NOT_FOUND` enables us to compare the error code `404` to `NOT_FOUND`?

Comment (c0derMo) [reply frontend/restInterfaces/BaseRestInterface.ts]: This is already known and being fixed

**Review by c0derMo — APPROVED** [review]
lgtm - well done

File: frontend/src/restInterfaces/AccountRestInterface.ts
Comment (Kr0nox) [inline frontend/src/restInterfaces/AccountRestInterface.ts]: Can you quickly explain why we have to have the option to return undefined?
Also for all other methods where this is present

**Review by Kr0nox — APPROVED** [review]

**Pull request merged**
